src/trainer_card.c

```diff
@@ -0,0 +1,1857 @@
+    trainerCard->hofDebutSeconds = playTime & 0xFF;
+    if ((playTime >> 16) > 999)
+    {
+        trainerCard->hofDebutHours = 999;
+        trainerCard->hofDebutMinutes = 59;
+        trainerCard->hofDebutSeconds = 59;
+    }
+
+    trainerCard->hasPokedex = FlagGet(FLAG_SYS_POKEDEX_GET);
+    trainerCard->caughtAllHoenn = HasAllHoennMons();
+    trainerCard->caughtMonsCount = GetCaughtMonsCount();
+
+    trainerCard->trainerId = (gSaveBlock2Ptr->playerTrainerId[1] << 8) | gSaveBlock2Ptr->playerTrainerId[0];
+
+    trainerCard->linkBattleWins = GetCappedGameStat(GAME_STAT_LINK_BATTLE_WINS, 9999);
+    trainerCard->linkBattleLosses = GetCappedGameStat(GAME_STAT_LINK_BATTLE_LOSSES, 9999);
+
+    trainerCard->pokemonTrades = GetCappedGameStat(GAME_STAT_POKEMON_TRADES, 0xFFFF);
+
+    trainerCard->money = GetMoney(&gSaveBlock1Ptr->money);
+
+    for (i = 0; i < 4; i++)
+        trainerCard->var_28[i] = gSaveBlock1Ptr->unk2BB0[i];
+
+    StringCopy(trainerCard->playerName, gSaveBlock2Ptr->playerName);
+
+    switch (cardType)
+    {
+    case CARD_TYPE_EMERALD:
+        trainerCard->battleTowerWins = 0;
+        trainerCard->battleTowerStraightWins = 0;
+    // Seems like GF got CARD_TYPE_FR and CARD_TYPE_RUBY wrong.
+    case CARD_TYPE_FR:
+        trainerCard->contestsWithFriends = GetCappedGameStat(GAME_STAT_WON_LINK_CONTEST, 999);
+        trainerCard->pokeblocksWithFriends = GetCappedGameStat(GAME_STAT_POKEBLOCKS_WITH_FRIENDS, 0xFFFF);
+        if (CountPlayerContestPaintings() > 4)
+            trainerCard->hasAllPaintings = TRUE;
+        trainerCard->stars = GetRubyTrainerStars(trainerCard);
+        break;
+    case CARD_TYPE_RUBY:
+        trainerCard->battleTowerWins = 0;
+        trainerCard->battleTowerStraightWins = 0;
+        trainerCard->contestsWithFriends = 0;
+        trainerCard->pokeblocksWithFriends = 0;
+        trainerCard->hasAllPaintings = 0;
+        trainerCard->stars = 0;
+        break;
+    }
+}
+
+static void sub_80C3020(struct TrainerCard *trainerCard)
+{
+    memset(trainerCard, 0, sizeof(struct TrainerCard));
+    trainerCard->version = VERSION_EMERALD;
+    SetPlayerCardData(trainerCard, CARD_TYPE_EMERALD);
+    trainerCard->hasAllSymbols = HasAllFrontierSymbols();
+    trainerCard->frontierBP = gSaveBlock2Ptr->frontier.field_EBA;
+    if (trainerCard->hasAllSymbols)
+        trainerCard->stars++;
+
+    if (trainerCard->gender == FEMALE)
+        trainerCard->var_4F = gUnknown_08329D54[(trainerCard->trainerId % 8) + 8];
+    else
+        trainerCard->var_4F = gUnknown_08329D54[trainerCard->trainerId % 8];
+}
+
+void TrainerCard_GenerateCardForPlayer(struct TrainerCard *trainerCard)
+{
+    memset(trainerCard, 0, 0x60);
+    trainerCard->version = VERSION_EMERALD;
+    SetPlayerCardData(trainerCard, CARD_TYPE_EMERALD);
+    trainerCard->var_3A = HasAllFrontierSymbols();
+    *((u16*)&trainerCard->berruCrushPoints) = gSaveBlock2Ptr->frontier.field_EBA;
```

Comment:
field_EBA might be a signed field which is causing the cast

Reply:
That's not it.